app/helpers/application_helper.rb

```diff
@@ -2,4 +2,84 @@ module ApplicationHelper
   def api_path
     "/api/content#{request.path}"
   end
+
+  def breadcrumbs
+    default_data = [{ title: "Home", url: "/" }]
+
+    return default_data if breadcrumbs_data.nil?
+
+    data = ordered_breadcrumbs_data
+
+    default_data + data.map { |breadcrumb| component_data(breadcrumb) }
+  end
+
+  def related_links
+    return [] if related_links_data.nil? || related_links_data.empty?
+
+    # partition child and parent links
+    parents = related_links_data.select { |item| !item.key?(:links) || item[:links].empty? }
+    children = related_links_data - parents
+    parents.each { |parent| parent[:children] = [] }
+
+    # assign children to their parents
+    children.each do |child|
+      parent = parents.find { |p| p[:content_id] == child[:links][:parent].first }
+      parent[:children] << child
+    end
+
+    # order parents and format for component rendering
+    ordered_parents(parents).map { |parent| component_data(parent) }
+  end
+
+private
+
+  def component_data(content_item)
+    output = {
+      title: content_item[:title],
+      url: content_item[:base_path]
+    }
+
+    if content_item.key?(:children)
+      output[:items] = content_item[:children].map { |child| component_data(child) }
+    end
+
+    output
+  end
+
+  def ordered_parents(parents)
+    ordered_breadcrumbs = ordered_breadcrumbs_data
+
+    parents.sort { |a, b| ordered_breadcrumbs.index(b) <=> ordered_breadcrumbs.index(a) }
+  end
+
+  def ordered_breadcrumbs_data
+    return [] if breadcrumbs_data.nil?
+
+    breadcrumbs_data.sort do |a, b|
+      case
+      when !b[:links] && a[:links]
+        1
+      when !a[:links] && b[:links]
+        -1
+      when b[:links].empty? && a[:links].key?(:parent)
+        1
+      when a[:links].empty? && b[:links].key?(:parent)
+        -1
+      when a[:links][:parent].first == b[:content_id]
+        1
+      when b[:links][:parent].first == a[:content_id]
+        -1
+      else
+        0
+      end
+    end
+  end
+
+  def breadcrumbs_data
+    @presenter.links[:parent] if @presenter.links && @presenter.links.any?
+  end
+
+  def related_links_data
+    @presenter.links[:related] if @presenter.links && @presenter.links.any?
+  end
 end
```

Comment:
I think this would be made simpler if the presenter encapsulated the breadcrumbs logic.

Right now, the ApplicationHelper needs to know quite a lot about the `links` data structure and business logic of how to order each breadcrumb. I wonder if this would be made clearer if it were handled in the presenter. I'd imagine it'd be easier to test, too. Would be happy to work on this together.

That being said, this works and you could also just ship it with a view to refactoring later.